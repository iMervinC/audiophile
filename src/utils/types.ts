import type { Db, MongoClient } from 'mongodb'
import type { NextApiRequest, NextApiResponse } from 'next/types'
import { Dispatch, SetStateAction } from 'react'

export interface Products {
  id: number
  slug: string
  name: string
  image: Image
  category: 'earphones' | 'headphones' | 'speakers'
  new: boolean
  price: number
  description: string
  features: string
  includes: Include[]
  gallery: Gallery
  others: Other[]
}

export interface Gallery {
  first: Image
  second: Image
  third: Image
}

export interface Image {
  mobile: string
  tablet: string
  desktop: string
}

export interface Include {
  quantity: number
  item: string
}

export interface Other {
  slug: string
  name: string
  image: Image
}

export interface MyApiRequest extends NextApiRequest {
  dbClient: MongoClient
  db: Db
}

export type MyController<T = any> = (
  req: MyApiRequest,
  res: NextApiResponse<T>
) => Promise<void>

export type Navigation = 'home' | 'headphones' | 'speakers' | 'earphones'

export interface TF {
  name?: string
  id?: string
  placeholder: string
  label: string
}

export interface RadioT {
  name?: string
  id?: string
  placeholder: string
  selected?: boolean
}

export interface CounterT {
  count: number
  setCount: Dispatch<SetStateAction<number>>
}
