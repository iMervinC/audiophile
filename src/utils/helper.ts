import axios from 'axios'
import { Products, CartProduct } from './types'

export const fetchByCategory = async (category: Products['category']) => {
  const { data } = await axios.get<Products[]>(
    `http://localhost:3000/api/products/${category}`
  )
  return data
}

export const totalCost = (cart: CartProduct[]) => {
  return (
    cart.length &&
    cart
      .map((item) => item.price! * item.quantity!)
      .reduce((acc, cur) => acc! + cur!)
  )
}

export const sortByNew = (a: Products, b: Products) => {
  return a.new === b.new ? 0 : a.new ? -1 : 1
}
