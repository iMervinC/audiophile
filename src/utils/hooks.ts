import { useEffect, useState } from 'react'

export const useGetScroll = () => {
  const [positionY, setPositionY] = useState(0)

  useEffect(() => {
    const pos = () => {
      setPositionY(window.scrollY)
    }

    window.addEventListener('scroll', pos)

    return () => window.removeEventListener('scroll', pos)
  }, [])

  return { scrollY: positionY }
}
