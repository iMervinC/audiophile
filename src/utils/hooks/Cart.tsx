import {
  useReducer,
  createContext,
  useContext,
  ReactNode,
  Reducer,
} from 'react'
import { CartProduct, InitialCartState, CartAction } from '../types'

interface DispatchActions {
  addToCart: (item: CartProduct) => void
  removeFromCart: (index: number) => void
  quantityChange: (type: 'Add' | 'Minus', index: number) => void
}

const CartStore = createContext<InitialCartState>({} as InitialCartState)
const CartDispatch = createContext<DispatchActions>({} as DispatchActions)

const initialState: InitialCartState = {
  cart: [],
}

const reducer: Reducer<InitialCartState, CartAction> = (state, action) => {
  switch (action.type) {
    case 'ADD':
      return { cart: [...state.cart, action.payload] }
    // === === === === === === === === ===
    case 'DELETE':
      const filltered = state.cart.filter((_, index) => index !== action.id)

      return {
        ...state,
        cart: [...filltered],
      }
    // === === === === === === === === ===
    case 'ADD_QUANTITY':
      state.cart[action.id].quantity! += 1

      return {
        ...state,
      }
    // === === === === === === === === ===
    case 'SUBTRACT_QUANTITY':
      state.cart[action.id].quantity = Math.max(
        0,
        (state.cart[action.id].quantity! -= 1)
      )

      return {
        ...state,
      }
    // === === === === === === === === ===

    default:
      return state
  }
}

export const CartProvider = ({ children }: { children: ReactNode }) => {
  const [state, dispatch] = useReducer(reducer, initialState)

  const addToCart = (item: CartProduct) => {
    dispatch({ type: 'ADD', payload: item })
  }

  const removeFromCart = (index: number) => {
    dispatch({ type: 'DELETE', id: index })
  }

  const quantityChange = (type: 'Add' | 'Minus', index: number) => {
    if (type === 'Add') {
      dispatch({ type: 'ADD_QUANTITY', id: index })
    } else if (type === 'Minus') {
      dispatch({ type: 'SUBTRACT_QUANTITY', id: index })
    }
  }

  return (
    <CartDispatch.Provider
      value={{ addToCart, removeFromCart, quantityChange }}
    >
      <CartStore.Provider value={state}>{children}</CartStore.Provider>
    </CartDispatch.Provider>
  )
}

export function useCartStore() {
  return useContext(CartStore)
}
export function useCartDispatch() {
  return useContext(CartDispatch)
}
