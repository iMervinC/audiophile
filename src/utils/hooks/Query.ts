import { useQuery } from 'react-query'
import axios from 'axios'
import { Products } from '../types'

export const useGetProducts = () => {
  const query = useQuery<Products[]>('Products', async () => {
    const { data } = await axios.get(`/api/products`)
    return data
  })

  return query
}

export const useGetProductsByCategories = (category: Products['category']) => {
  const query = useQuery<Products[]>(['Products', category], async () => {
    const { data } = await axios.get(`/api/products/${category}`)
    return data
  })

  return query
}
