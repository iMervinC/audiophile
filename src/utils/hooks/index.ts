export * from './GetScroll'
export * from './Query'
export * from './usePrefetchImmediately'
// export * from './Cart'
