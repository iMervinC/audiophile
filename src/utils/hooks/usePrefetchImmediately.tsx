import { useEffect } from 'react'
import { apiSlice } from '@/features/store/storeApiSlice'
import { useAppDispatch } from '@/app/hooks'

export type PrefetchOptions =
  | { force?: boolean }
  | {
      ifOlderThan?: false | number
    }
type EndpointNames = keyof typeof apiSlice.endpoints

export function usePrefetchImmediately<T extends EndpointNames>(
  endpoint: T,
  arg: any,
  options: PrefetchOptions = {}
) {
  const dispatch = useAppDispatch()
  useEffect(() => {
    dispatch(apiSlice.util.prefetch(endpoint, arg, options))
  }, [])
}
