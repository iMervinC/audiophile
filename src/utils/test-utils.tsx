import { render, RenderOptions, screen } from '@testing-library/react'
import { FC, ReactElement } from 'react'
import { CartStore, CartDispatch, CartProvider } from '@/utils/hooks'

const AllProviders: FC = ({ children }) => {
  const storeProps = {
    value: {
      cart: [
        {
          image: {
            desktop: '/assets/product-zx9-speaker/desktop/image-product.jpg',
            mobile: '/assets/product-zx9-speaker/mobile/image-product.jpg',
            tablet: '/assets/product-zx9-speaker/tablet/image-product.jpg',
          },

          name: 'ZX9 Speaker',
          price: 4500,
          quantity: 1,
          slug: 'zx9-speaker',
        },
      ],
    },
  }

  const dispatchProps = {
    value: {
      addToCart: jest.fn(),
      removeFromCart: jest.fn(),
      clearCart: jest.fn(),
      updateItem: jest.fn(),
    },
  }

  return (
    <CartDispatch.Provider {...dispatchProps}>
      <CartStore.Provider {...storeProps}>{children}</CartStore.Provider>
    </CartDispatch.Provider>
  )
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllProviders, ...options })

export * from '@testing-library/react'
export { customRender as render }
