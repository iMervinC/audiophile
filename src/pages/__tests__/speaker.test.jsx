import { render, screen } from '@testing-library/react'
import Speakers from '../speakers'
import { useGetProductsByCategories } from '@/utils/hooks'

const data = [
  {
    id: 1,
    slug: 'xx59-headphones',
    name: 'XX59 Headphones',
    category: 'headphones',
    new: true,
    image: {
      desktop: '/home/desktop/image-earphones-yx1.jpg',
    },
  },
  {
    id: 2,
    slug: 'asd-headphones',
    name: 'asd Headphones',
    category: 'headphones',
    new: false,
    image: {
      desktop: '/home/desktop/image-earphones-yx1.jpg',
    },
  },
]

jest.mock('@/utils/hooks', () => ({
  useGetProductsByCategories: jest.fn(),
}))

describe('Earphone', () => {
  let page

  beforeEach(() => {
    useGetProductsByCategories.mockImplementation(() => ({ data }))
    page = render(<Speakers />)
  })

  test('<Earphone/> Snapshot', () => {
    expect(page.container).toMatchSnapshot()
  })

  test('Heading', () => {
    expect(screen.getByTestId('categoryTitle')).toHaveTextContent('Speakers')
  })

  test('Categories Component', () => {
    expect(screen.getByTestId('Categories')).toBeTruthy()
  })

  test('About Component', () => {
    expect(screen.getByTestId('About')).toBeTruthy()
  })

  test('Products list', () => {
    expect(useGetProductsByCategories).toHaveBeenCalledWith(
      'speakers',
      undefined
    )
    data.forEach((item) => {
      expect(screen.getByText(new RegExp(item.name))).toBeInTheDocument
    })
  })

  test('Loading', () => {
    useGetProductsByCategories.mockImplementation(() => ({ isLoading: true }))
    render(<Speakers />)

    expect(screen.queryByTestId('Loading')).toBeInTheDocument()
  })

  test('Error', () => {
    expect(screen.queryByTestId('Error')).not.toBeInTheDocument()

    useGetProductsByCategories.mockImplementation(() => ({ isError: true }))
    render(<Speakers />)

    expect(screen.queryByTestId('Error')).toBeInTheDocument()
  })
})
