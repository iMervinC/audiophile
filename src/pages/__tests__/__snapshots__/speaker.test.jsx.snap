// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Earphone <Earphone/> Snapshot 1`] = `
<div>
  <main>
    <div
      class="bg-black pb-24 text-center pt-[212px] "
      data-testid="categoryTitle"
    >
      <h1
        class="text-white text-4xl md:title"
      >
        Speakers
      </h1>
    </div>
    <ul
      class="space-y-20 my-28 max-w-[1150px] container"
      data-testid="Products list"
    >
      <div
        class="preview-card md:flex-row"
        role="listitem"
      >
        <div
          class="w-full bg-grey text-center rounded-lg md:mr-20"
        >
          <div
            style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
          >
            <div
              style="box-sizing: border-box; display: block; max-width: 100%;"
            >
              <img
                alt=""
                aria-hidden="true"
                role="presentation"
                src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNTAwIiBoZWlnaHQ9IjUwMCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
                style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
              />
            </div>
            <noscript />
            <img
              alt="xx59-headphones"
              class="w-44"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: fill;"
            />
          </div>
        </div>
        <div
          class="flex flex-col justify-center space-y-9 w-full items-center md:items-start"
        >
          <div>
            <p
              class="overline text-main "
            >
              New Product
            </p>
            <h2
              class="text-4xl"
            >
              XX59 Headphones
            </h2>
          </div>
          <p />
          <button
            class="undefined bg-main text-white hover:bg-main-second w-40 h-12 transition-colors sub-title"
            title="Buttonsee-product"
            type="submit"
          >
            See Product
          </button>
        </div>
      </div>
      <div
        class="preview-card md:flex-row-reverse"
        role="listitem"
      >
        <div
          class="w-full bg-grey text-center rounded-lg md:ml-20"
        >
          <div
            style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
          >
            <div
              style="box-sizing: border-box; display: block; max-width: 100%;"
            >
              <img
                alt=""
                aria-hidden="true"
                role="presentation"
                src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNTAwIiBoZWlnaHQ9IjUwMCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
                style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
              />
            </div>
            <noscript />
            <img
              alt="asd-headphones"
              class="w-44"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: fill;"
            />
          </div>
        </div>
        <div
          class="flex flex-col justify-center space-y-9 w-full items-center md:items-start"
        >
          <div>
            <h2
              class="text-4xl"
            >
              asd Headphones
            </h2>
          </div>
          <p />
          <button
            class="undefined bg-main text-white hover:bg-main-second w-40 h-12 transition-colors sub-title"
            title="Buttonsee-product"
            type="submit"
          >
            See Product
          </button>
        </div>
      </div>
    </ul>
    <ul
      class="container flex flex-col justify-between items-center max-w-[1150px] my-28 space-y-20  sm:flex-row sm:space-y-0 sm:space-x-5 "
      data-testid="Categories"
    >
      <li
        class="text-center sm:w-[350px] w-10/12  relative undefined"
      >
        <div
          style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
        >
          <div
            style="box-sizing: border-box; display: block; max-width: 100%;"
          >
            <img
              alt=""
              aria-hidden="true"
              role="presentation"
              src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjUwIiBoZWlnaHQ9IjIyNSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
              style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
            />
          </div>
          <noscript />
          <img
            class="z-20"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
        </div>
        <div
          class="bg-grey rounded-2xl pt-32 pb-7 absolute w-full top-16 z-10 space-y-5 sm:top-14"
        >
          <h6
            class="uppercase"
          >
            headphones
          </h6>
          <button
            class="undefined sub-title w-[fit-content] h-[fit-content] text-grey-shop hover:text-main"
            title="Buttonheadphones"
          >
            Shop
            <svg
              class="inline ml-3"
              height="12"
              width="8"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M1.322 1l5 5-5 5"
                fill="none"
                fill-rule="evenodd"
                stroke="#D87D4A"
                stroke-width="2"
              />
            </svg>
          </button>
        </div>
      </li>
      <li
        class="text-center sm:w-[350px] w-10/12  relative undefined"
      >
        <div
          style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
        >
          <div
            style="box-sizing: border-box; display: block; max-width: 100%;"
          >
            <img
              alt=""
              aria-hidden="true"
              role="presentation"
              src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjUwIiBoZWlnaHQ9IjIyNSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
              style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
            />
          </div>
          <noscript />
          <img
            class="z-20"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
        </div>
        <div
          class="bg-grey rounded-2xl pt-32 pb-7 absolute w-full top-16 z-10 space-y-5 sm:top-14"
        >
          <h6
            class="uppercase"
          >
            speakers
          </h6>
          <button
            class="undefined sub-title w-[fit-content] h-[fit-content] text-grey-shop hover:text-main"
            title="Buttonspeakers"
          >
            Shop
            <svg
              class="inline ml-3"
              height="12"
              width="8"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M1.322 1l5 5-5 5"
                fill="none"
                fill-rule="evenodd"
                stroke="#D87D4A"
                stroke-width="2"
              />
            </svg>
          </button>
        </div>
      </li>
      <li
        class="text-center sm:w-[350px] w-10/12  relative undefined"
      >
        <div
          style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
        >
          <div
            style="box-sizing: border-box; display: block; max-width: 100%;"
          >
            <img
              alt=""
              aria-hidden="true"
              role="presentation"
              src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjUwIiBoZWlnaHQ9IjIyNSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
              style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
            />
          </div>
          <noscript />
          <img
            class="z-20"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
        </div>
        <div
          class="bg-grey rounded-2xl pt-32 pb-7 absolute w-full top-16 z-10 space-y-5 sm:top-14"
        >
          <h6
            class="uppercase"
          >
            earphones
          </h6>
          <button
            class="undefined sub-title w-[fit-content] h-[fit-content] text-grey-shop hover:text-main"
            title="Buttonearphones"
          >
            Shop
            <svg
              class="inline ml-3"
              height="12"
              width="8"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M1.322 1l5 5-5 5"
                fill="none"
                fill-rule="evenodd"
                stroke="#D87D4A"
                stroke-width="2"
              />
            </svg>
          </button>
        </div>
      </li>
    </ul>
    <div
      class="flex flex-col-reverse container px-9 my-60 space-y-10 max-w-[1150px] gap-[50px] h-auto lg:flex-row lg:space-x-10 md:space-y-0 lg:h-[588px] lg:px-0 "
      data-testid="About"
    >
      <div
        class="flex flex-col text-center lg:text-left lg:justify-center h-full md:w-full"
      >
        <h3
          class="title text-4xl mb-10"
        >
          Bringing you the 
          <span
            class="text-main"
          >
            best
          </span>
           audio gear
        </h3>
        <p>
          Located at the heart of New York City, Audiophile is the premier store for high end headphones, earphones, speakers, and audio accessories. We have a large showroom and luxury demonstration rooms available for you to browse and experience a wide range of our products. Stop by our store to meet some of the fantastic people who make Audiophile the best place to buy your portable audio equipment.
        </p>
      </div>
      <div
        class="bg-about-mobile sm:bg-about-tablet lg:bg-about bg-no-repeat bg-cover bg-center h-[300px] w-full rounded-xl lg:h-auto"
      />
    </div>
  </main>
</div>
`;
